x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ApiKeyAggregationsBuilder.java

```diff
@@ -0,0 +1,146 @@
+            for (PipelineAggregationBuilder pipelineAggregator : copiedAggsBuilder.getPipelineAggregations()) {
+                throw new IllegalArgumentException("Unsupported pipeline aggregation of type [" + pipelineAggregator.getType() + "]");
+            }
+            if (copiedAggsBuilder instanceof ValuesSourceAggregationBuilder<?> valuesSourceAggregationBuilder) {
+                if (valuesSourceAggregationBuilder instanceof TermsAggregationBuilder == false
+                    && valuesSourceAggregationBuilder instanceof RangeAggregationBuilder == false
+                    && valuesSourceAggregationBuilder instanceof DateRangeAggregationBuilder == false
+                    && valuesSourceAggregationBuilder instanceof MissingAggregationBuilder == false
+                    && valuesSourceAggregationBuilder instanceof CardinalityAggregationBuilder == false
+                    && valuesSourceAggregationBuilder instanceof ValueCountAggregationBuilder == false) {
+                    throw new IllegalArgumentException(
+                        "Unsupported API Keys agg [" + copiedAggsBuilder.getName() + "] of type [" + copiedAggsBuilder.getType() + "]"
+                    );
+                }
+                // scripts are not currently supported because it's harder to restrict and rename the doc fields the script has access to
+                if (valuesSourceAggregationBuilder.script() != null) {
+                    throw new IllegalArgumentException("Unsupported script value source for [" + copiedAggsBuilder.getName() + "] agg");
+                }
+                // the user-facing field names are different from the index mapping field names of API Key docs
+                String translatedFieldName = ApiKeyFieldNameTranslators.translate(valuesSourceAggregationBuilder.field());
+                valuesSourceAggregationBuilder.field(translatedFieldName);
+                fieldNameVisitor.accept(translatedFieldName);
+                return valuesSourceAggregationBuilder;
+            } else if (copiedAggsBuilder instanceof CompositeAggregationBuilder compositeAggregationBuilder) {
+                for (CompositeValuesSourceBuilder<?> valueSource : compositeAggregationBuilder.sources()) {
+                    if (valueSource.script() != null) {
+                        throw new IllegalArgumentException(
+                            "Unsupported script value source for ["
+                                + valueSource.name()
+                                + "] of composite agg ["
+                                + compositeAggregationBuilder.getName()
+                                + "]"
+                        );
+                    }
+                    String translatedFieldName = ApiKeyFieldNameTranslators.translate(valueSource.field());
+                    valueSource.field(translatedFieldName);
+                    fieldNameVisitor.accept(translatedFieldName);
```

Comment:
This just occurred to me: (for a follow up, not this PR!) I think it could be less error prone to extend the `translate` signature to accept the field name visitor or create a new method `translateWithVisitor()` to avoid missing a call to `fieldNameVisitor.accept`.

Reply:
I thought about it too. I'm not sure it's a great interface for the `#translate` method...